### ISSUE TITLE
X25519 keys endianness incompatible with mbedTLS

### ISSUE THREAD

**Issue by Jongy** [pyca/cryptography#4981]
I toyed with an X25519 demo of mbedTLS <-> cryptography.

**TL;DR** by reversing the output of `public_bytes()` before passing it to mbedTLS, by reversing the public part mbedTLS gives me before passing it to cryptography and by reversing the shared secret of either side, everything works.
_Why?_

In mbedTLS,  `mbedtls_mpi_write_binary` is used to "export" the bytes of the public part, as well as the final secret itself.
`mbedtls_mpi_write_binary` clearly states `Export X into unsigned binary data, big endian.`

In cryptography, when using `X25519PublicKey.public_bytes()`, it uses the OpenSSL backend (apparently hardcoded, at least as of ca723c42d0e48c4a17aaed2505c2a5be80e1165f) which ends up calling `EVP_PKEY_get1_tls_encodedpoint`.

I dived deeper and although I can't say I found the exact function doing the conversion (it's all rather messed up with lots of function tables), I did find a few functions implementing this conversion (perhaps not for Curve25519 though) and they all used the bignum API, by using `BN_bn2bin` which is apparently big endian as well, as seen [here](https://www.openssl.org/docs/man1.0.2/man3/BN_bn2bin.html).

I also tried with other curves (e.g SECP256R1) which allow accessing the point's X and Y via `public_numbers().X`. Then I used `public_bytes(Encoding.X962, PublicFormat.CompressedPoint)[1:]` and the resulting 32 bytes are indeed big endian: `public_numbers().X == int.from_bytes(public..., byteorder='big')`.

I also found [this](https://forums.mbed.com/t/incompatibility-between-x25519-key-created-with-openssl/3078) forum post in mbedTLS, explaining the same issue.

I think it should be documented in the X25519 docs, but it should be checked deeply to realize what's the root cause.

**Comment by reaperhulk** [comment]
X25519 byte level encoding is defined as little-endian in [RFC 7748](https://www.ietf.org/rfc/rfc7748.txt). This is in contrast with many other things, which are typically encoded in "network byte order" aka big-endian. mbedTLS is wrong here if they want to be able to interop easily. We would, however, be happy to take a PR that explicitly notes that this encoding is little-endian for cases like this.

**Comment by Jongy** [comment]
Whatd'ya say. I didn't even think of checking a definition document, I was sure it's about implementation incompatibility which I could find in the code.  I'll make a PR stating the ensiannes for both Curve25519 and Curve448.

**Comment by stepheny** [comment]
`mbedtls_mpi_write_binary` and `mbedtls_mpi_read_binary` are just bignum read/write methods. They're designed to read/write in big endian. Nothing is wrong with eigher side.
There are `mbedtls_mpi_write_binary_le` and `mbedtls_mpi_read_binary_le` work in little endian. I believe they exists in development branch. I'm not sure whether they have reached a stable release.

**Comment by alex** [comment]
After discussing with @reaperhulk, we're going to close this -- since every element of this module follows the RFC, calling one specific element of the encoding doesn't make sense.
